### DISCUSSION TITLE
how to recognize RCP usb port when there are multiple usbs

### DISCUSSION THREAD

**Discussion by Leo2442926161** [openthread/openthread#9623] [Q&A]
Hi all，
would you recommend how to recognize RCP usb port when there are multiple usbs in linux? e.g. ttyUSB0, ttyUSB1 and ttyUSB2, how application knows which usb port is right RCP port and update it to otbr configuration?  Thanks.

**Comment by jwhui** [comment]
Take a look at [`libudev`](https://manpages.ubuntu.com/manpages/mantic/en/man3/libudev.3.html).

**Reply by Leo2442926161** [reply]
Thanks @jwhui .

if there are multiple usb dongles, how to make the udev rules?  
for example:
ttyUSB0: zigbee dongle
ttyUSB1: openthread dongle
ttyUSB2: zwave dongle

how can my code knows the ttyUSB1 is the rcp?  if the usb number changed, how udev filter the right rcp as soon as possible?

**Reply by jwhui** [reply]
The dongles should present a different (vendor id, product id) combination. That should allow you to differentiate.

**Comment by Leo2442926161** [comment]
> The dongles should present a different (vendor id, product id) combination. That should allow you to differentiate.

Thanks again, what if the dongles used same USB2UART chipset(like cp201x), just the radio nuance.

**Reply by jwhui** [reply]
That depends on your chip - a good one will allow you to configure the vendor id / product id. For example: https://ftdichip.com/wp-content/uploads/2020/07/AN_124_User_Guide_For_FT_PROG.pdf

**Reply by Leo2442926161** [reply]
@jwhui  thanks, configure the USB2UART chip is not convenient in facrory.

where can I find rcp uart protocol? from my perspective, seems the host to detect the dongle via uart protocol is more sensible.

**Reply by jwhui** [reply]
See: https://github.com/openthread/openthread/blob/main/src/lib/spinel/spinel.h

OpenThread will exit if it cannot communicate properly with the designated RCP. You could try simply having a script that tries different device paths when that failure occurs.

**Reply by Leo2442926161** [reply]
Thanks very much @jwhui
